pkg/cached/cached.go

```diff
@@ -2,368 +2,598 @@ package cached
+	if notMounted {
+		logger.Debug(ctx, "checking if base lv is formatted", key.LV.Field(lv))
+		var format string
+		format, err = exec.Output(ctx, "blkid", "-o", "value", "-s", "TYPE", lvDevice)
+		if err != nil && !strings.Contains(err.Error(), "exit status 2") {
+			return fmt.Errorf("failed to get filesystem type of base device %s: %w", lvDevice, err)
+		}
+
+		if format != c.BaseLVFormat {
+			logger.Info(ctx, "formatting base lv", key.LV.Field(lv), zap.String("expected", c.BaseLVFormat), zap.String("actual", format))
+			if err := exec.Run(ctx, "mkfs."+c.BaseLVFormat, FormatOptions(lvDevice, c.BaseLVFormat)...); err != nil {
+				return fmt.Errorf("failed to format base lv %s: %w", lvDevice, err)
+			}
+		}
 
-	trace.SpanFromContext(ctx).SetAttributes(
-		key.TargetPath.Attribute(targetPath),
-		key.VolumePath.Attribute(volumePath),
-		key.UpperDir.Attribute(upperDir),
-		key.WorkDir.Attribute(workDir),
-	)
+		if err = os.MkdirAll(c.BaseLVMountPoint, 0o775); err != nil {
+			return fmt.Errorf("failed to mkdir %s: %w", c.BaseLVMountPoint, err)
+		}
 
-	// Check if the volume path exists
-	_, err := os.Stat(volumePath)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return &csi.NodeUnpublishVolumeResponse{}, nil // Nothing for us to do
+		logger.Info(ctx, "mounting base lv", key.LV.Field(lv), key.Path.Field(c.BaseLVMountPoint))
+		if err := mounter.Mount(lvDevice, c.BaseLVMountPoint, c.BaseLVFormat, MountOptions(c.BaseLVFormat)); err != nil {
+			return fmt.Errorf("failed to mount %s to %s: %w", lvDevice, c.BaseLVMountPoint, err)
+		}
+
+		// ensure the base lv is unmounted when the function returns
+		defer func() {
+			if notMounted, _ := mounter.IsLikelyNotMountPoint(c.BaseLVMountPoint); !notMounted {
+				if err := mounter.Unmount(c.BaseLVMountPoint); err != nil {
+					logger.Error(ctx, "failed to unmount base lv", zap.Error(err))
+				}
+			}
+		}()
+
+		// Clean up lost+found directory, it's not needed and confusing.
+		if err := os.RemoveAll(filepath.Join(c.BaseLVMountPoint, "lost+found")); err != nil {
+			return fmt.Errorf("failed to delete %s: %w", filepath.Join(c.BaseLVMountPoint, "lost+found"), err)
 		}
-		return nil, fmt.Errorf("failed to stat volume path %s: %w", volumePath, err)
 	}
 
-	// Check if the overlay is mounted
-	err = execCommand("mountpoint", "-q", targetPath)
-	if err == nil {
-		// The overlay is mounted, so we need to unmount it
-		if err := execCommand("umount", targetPath); err != nil {
-			return nil, fmt.Errorf("failed to unmount overlay at %s: %w", targetPath, err)
+	cacheRootDir := filepath.Join(c.BaseLVMountPoint, "dl_cache")
+	cacheVersions := client.ReadCacheVersionFile(cacheRootDir)
+
+	if cacheVersion == -1 || !slices.Contains(cacheVersions, cacheVersion) {
+		logger.Info(ctx, "downloading cache", key.CacheVersion.Field(cacheVersion))
+		startTime := time.Now()
+
+		var cachedCount uint32
+		cacheVersion, cachedCount, err = c.Client.GetCache(ctx, cacheRootDir, cacheVersion)
```

Comment:
Do we still even want this functionality? I guess its convenient in tests? Maybe we want a flag that disallows this or something though so it doesnt happen in prod by accident?